Добавить FastAPI (#21)

Rey-Lane opened the pull request:
Т.к. в качестве фреймворка для бэкенда был выбран фастапи, необходимо добавить его в проект и убедиться, что всё работает.

Чтобы убедиться, что фастапи установлен правильно и работает, необходимо будет создать простой роутер, и написать к нему тест.

В рамках данной задачи необходимо:
- убрать уже не нужные `src/ping.py` и `tests/test_ping.py`
- добавить `fastapi` в `pyproject.toml`, обновить `poetry.lock` и установить `fastapi` локально
- создать инстанс приложения в `src/app.py`
- добавить роут `GET /health` в `src/health/routes.py` и подключить роутер к приложению
- добавить тест для этого роута в `tests/test_health.py`

birthdaysgift commented on pyproject.toml:
```diff
@@ -15,12 +15,14 @@ version = "0.0.0"
 
 darglint = "1.8.1"  # Checks whether a docstring's description matches the actual function/method implementation
 dlint = "0.14.0"  # Tool for encouraging best coding practices and helping ensure Python code is secure.
+fastapi = {extras = ["all"], version = "0.95.0"}  # We use extras = ["all"] because fastapi TestClient requers some additional dependencies
```
Чтобы коммент не выходил за 120 символов его можно немного перефразировать:
```python
# We use "all" here because fastapi TestClient doesn't work without it
```

birthdaysgift commented on pyproject.toml:
```diff
@@ -15,12 +15,14 @@ version = "0.0.0"
 [tool.poetry.dependencies]
 
 python = "~3.11"
+fastapi = "0.95.0"
```
Да, мне бы тоже хотелось оставить пайтон на самом верху и не сортировать его алфавитно вместе с библиотеками проекта, но давай тогда отделим его от остальных зависимостей визуально, тип такого:
```toml
[tool.poetry.dependencies]

python = "~3.11"

fastapi = "0.95.0"
psycopg-binary = "1.2.3"
sqlalchemy = "4.5.6"
```

birthdaysgift commented on tests/test_service/test_healthcheck.py:
```diff
@@ -0,0 +1,12 @@
+client = TestClient(app)
+
+
+def test_healthcheck():
```
Насколько я знаю, тесты тоже нужно сделать асинхронными. Пока не совсем понимаю зачем это делать, но в доке fastapi есть такой [раздел](https://fastapi.tiangolo.com/advanced/async-tests/).

Возможно нам еще пригодится [pytest-asyncio](https://pypi.org/project/pytest-asyncio/), но пока не уверен...

UPD: pytest-asyncio не нужен, потому что при установке `fastapi[all]` ставится `anyio`, и, кажется, он позволяет делать асинхронные тесты через `@pytest.mark.anyio`

PS: Судя по всему `TestClient` также придётся заменить на `AsyncClient` из `httpx`.

birthdaysgift commented on src/app.py:
```diff
@@ -0,0 +1,7 @@
+"""Main fastAPI application."""
+from fastapi import FastAPI
```
Предлагаю договориться выделять блок всех импортов одной пустой строкой сверху и двумя снизу, т.е.:
```python
"""Docstring."""

from fastapi import FastAPI

from src.routes import router


app = FastAPI()
app.include_router()
```
Одна пустая строка перед списком импортов будет визуально отделять их от докстринга. А две строки после импортов будут отделять их от остального кода. Если возникает вопрос, почему именно две строки, а не одна, то это для того чтобы отличать разделитель перед кодом от разделителя между группами импортов. Группы импортов отделяются друг от друга одной линией, а импорты от кода - двумя.

Так же предлагаю сразу договориться о расположении и содержании блока `if TYPE_CHECKING:`. Предлагаю писать этот блок после всех импортов, и отделять его одной строчкой сверху. Импорты внутри блока писать по тем же правилам, что и снаружи, т.е. сначала системные модули, потом сторонние, и потом модули приложения. Каждая группа отделяется одной строкой, все импорты отсортированы в алфавитном порядке.

Таким образом полный пример будет выглядеть как-то так:

```python
"""Docstring."""

from typing import TYPE_CHECKING

from fastapi import FastAPI

from src.routes import router

if TYPE_CHECKING:
    from typing import Iterable

    from fastapi import HTTPException

    from src.models import User
    from src.utils import pretty_print


app = FastAPI()
app.include_router()
```

К сожалению линтеры, это не проверяют, поэтому придётся об этом договориться и следить за этим своими силами (ну либо написать плагин флейку и перейти на флейк, если есть желание :smile: ).

Если апрувим такой подход, то могу зафиксировать это в вики бэкенда.

Rey-Lane replied:
Тут ruff ругался на двойной отступ после импортов, поэтому здесь не стал делать

birthdaysgift commented on src/app.py:
```diff
@@ -0,0 +1,7 @@
+"""Main fastAPI application."""
```
Предлагаю сделать либо `fastapi`, либо `FastAPI`.

А то какой-то странный подоконник получается : )

birthdaysgift commented:
Также есть ряд вопросов, которые я не стал писать в ревью, т.к. я пока сам до конца не уверен, как сделать лучше. Над этим надо подумать и принять решение.

1. Т.к. судя по всему healthcheck роут не будет удалён, а станет частью системы, и будет действительно использоваться для проверки состояния приложения, то надо бы решить куда его положить.

   Пока что у нас два варианта - тот который сейчас, т.е. в папке `service` ("обслуживание", в смысле техническое) или в своей папке `healthcheck`. Возможно, появится третий вариант и он будет ещё лучше (спойлер - он появился в пятом пункте этого текста).

2. Меня смущает использование названия проекта (wisher) непосредственно в коде приложения, т.е. в `src`. Боюсь что это вызовет трудности при ребрендинге. Хотя, конечно, понятно, что ребрендинг и так вызовет трудности в случае чего.

   Тем не менее, кажется, что можно везде слово "wisher" заменить на "app", без какой-либо потери смысла.

3. Хотелось бы упростить нейминг роутеров. Мне кажется, что все роутеры лучше просто называть `router`.

   Кажется нет смысла писать, допустим, `user_router` если этот роутер и так лежит в `src.user.routes`. Разделение кода по пакетам привязанным к фиче как раз-таки позволяет нам в некоторых моментах упрощать нейминг. Например, если схема для кредов юзера лежит в `src.user.schemas` то её можно назвать просто `Credentials`, а не `UserCredentialsSchema`.

4. Многочасовые совещания с командой фронтенда выявили, что с эндпойнтами, которые возвращают просто строчку не очень удобно работать (это я про healthcheck). Правда в интернете я так и не нашёл какого-то единого мнения по этому вопросу.

   Но т.к. нам уже довелось поработать без энвелопа на других проектах (ну вы понимаете какой проект я имею в виду), то предлагаю в этот раз попробовать заворачивать все респонзы в какую-то единую структуру. Только вот в какую именно структуру - опять же вопрос открытый...

   <details><summary>Тут меня чет вообще далеко понесло, лучше даже не открывать...</summary>
   <p>

   <details><summary>Не уверен, что там есть что-то полезное.</summary>
   <p>

   <details><summary>Я предупредил...</summary>
   <p>

   На сколько я понял эта структура обычно на верхнем уровне передаёт всякого рода метаданные, а фактические данные представленны в каком-то определённом аттрибуте, например:
   ```python
   {
       "data": {
           ...  # здесь будут запрошенные данные
       },

       # всё остальное это метаданные, например
       "quantity": {  # количество
           "current": 3,  # сколько вернули
           "total": 7239  # сколько всего
       }
   }
   ```

   Но т.к. вряд ли у нас получится придумать такую схему метаданных, которая подходила абсолютно для всех ответов, поэтому можно попробовать пойти другим путём, и предоставить возможность возвращать разные метаданные для разных эндпойнтов, т.е.:
   ```python
   {
       "data": {
           ...  # запрошенные данные
       },
       "meta": {
           ...  # метаданные
       }
   }

   ```
   Таким образом мы сможем зафиксировать общую схему энвелопа - любой ответ должен возвращать объект с двумя полями - "data" и "meta". Для разных эндпойнтов мета может быть разная. Таким образом в коде это будут разные схемы, типо:
   ```python
   # src/users/schemas.py


   class UserData:
       """User info returned."""
       ...  # какая-то схема для юзера


   class UserMeta:
       """User info metadata."""
       ...  # какая-то схема для метаданных по юзеру
            # (хотя это странно конечно, как тип метаданных может быть привязан к конкретной сущности типа юзера)
   ```
   Так же можно завести общие схемы для данных и метаданных, например:
   ```python
   # src/shared/schemas.py


   class PaginationData(GenericSchema, Generic[T]):
       """Generic schema for paginated data in response."""

       data: Dict[str, List[T]]


   class PaginationMeta:
       """Generic schema for paginated metadata."""

       limit: int
       offset: int


   class ValueData(GenericSchema, Generic[T]):
       """Generic schema for value data in response."""
       data: Dict[str, T]


   class ValueMeta(GenericSchema, Generic[T]):
       """Generic schema for value metadata."""

       meta: Dict[str, T]  # я хз че тут может быть...
   ```
   И так же у нас будет общая схема тела ответа (энвелопа):
   ```python
   # src/shared/schemas.py


   DataSchema = TypeVar("DataSchema", bound=dict)
   MetaSchema = TypeVar("MetaSchema", bound=dict)


   class Response(GenericSchema, Generic[DataSchema], Generic[MetaSchema]):
       """Generic class for all responses."""
       data: DataSchema
       meta: MetaSchema
   ```

   _Всё, тут конец. Да вывода нет. Че там за дженерик схемы и надо ли оно нам, я сам не не знаю..._

   </p>
   </details> 

   </p>
   </details>

   </p>
   </details>

4. Возможно будет проще если переделать роут `/healthcheck` в сущность, типа `/health`. Тогда запрос `GET /health` будет восприниматься однозначно, как "получить инфу о здоровье". В качестве респонза можно отдать что-то типа `{"status": "OK"}`. Т.е. как будто у нас есть сущность "здоровье", у которой одно поле `status`.

   Таким образом мы сделаем эндпойнт более "RESTful" и при этом избавимся от респонза в виде строки.

   Так же у нас сейчас есть небольшая чехарда с написанием слова healthcheck. Тут буквально в трех соседних строчках, три разных написания:

   ![image](https://user-images.githubusercontent.com/35929293/231914334-81c047ce-693a-4a1b-972c-82283c7f86eb.png)

   Соответственно, если мы переделаем эндпойнт на `/health`, такой проблемы не будет.

   Заодно может решиться проблема из пункта **1**. Поскольку теперь `health` это сущность, то можно и положить её в свою отдельную папку - `src/health`. Конечно, немного смущает, что эта сущность не доменная, но хоть это и так, почему она не имеет права на свою отдельную директорию. Иначе получится странно, если доменные сущности будут красиво лежать по папочкам, а все остальные сущности будут свалены в кучу. Несправедливо как-то...

6. Ещё смутил коммит-месседж. Вроде как практика копирования текста тикета в коммит месседж неплохо себя зарекомендовала... В принципе, если это неудобно делать руками, то я думаю можно написать пайтон скриптик куда-нибудь в `src/envs/dev/scripts/git`, который будет из названия ветки парсить номер тикета, идти на гитхаб, копировать оттуда тайтл и дескрипшн, сквашить коммиты и вписывать правильный коммит-месседж.

Commit:
21: Добавить FastAPI

Т.к. в качестве фреймворка для бэкенда был выбран фастапи, необходимо добавить его в проект и убедиться, что всё работает.

Чтобы убедиться, что фастапи установлен правильно и работает, необходимо будет создать простой роутер, и написать к нему тест.

В рамках данной задачи необходимо:
- убрать уже не нужные `src/ping.py` и `tests/test_ping.py`
- добавить `fastapi` в `pyproject.toml`, обновить `poetry.lock` и установить `fastapi` локально
- создать инстанс приложения в `src/app.py`
- добавить роут `GET /health` в `src/health/routes.py` и подключить роутер к приложению
- добавить тест для этого роута в `tests/test_health.py`

birthdaysgift commented:
Коммит, на который ссылается пулл реквест не существует, так как мы меняли соглашение по именованию коммитов и ребейзили `develop` ветку. Подробнее об этом можно почитать в [обсуджении](https://github.com/week-password/wisher-backend/discussions/3#discussioncomment-5640598).

Хэш нового коммита, содержащего точную копию изменений внесённых этим пулл-реквестом: 847a4d9
